core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAllocator.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.metadata.InternalNode;
+
+import java.io.Closeable;
+
+public interface NodeAllocator
```

Comment:
This interface serves a similar purpose as the already existing `NodeSelector`. Should we refactor what we already have instead of adding a new entity?

Reply:
I was also thinking about it. Though it seems like it needs more thought and a larger refactor. I would prefer to get this done in a separate PR as this PR is already quite huge.

Reply:
Yeah, let's have it in a follow up. I didn't put a lot of though into a design of this interface with an assumption that a better interface could be designed when somebody starts working specifically on resource management.